drivers/tty/serial/8250/8250_ni.c

```diff
@@ -277,74 +277,78 @@ static int ni16550_probe(struct platform_device *pdev)
 {
 	const struct ni16550_device_info *info;
 	struct device *dev = &pdev->dev;
-	struct uart_8250_port uart = {};
 	unsigned int txfifosz, rxfifosz;
-	unsigned int prescaler;
+	struct uart_8250_port *uart;
 	struct ni16550_data *data;
+	unsigned int prescaler;
 	const char *portmode;
 	bool rs232_property;
 	int ret;
 
+	uart = devm_kzalloc(dev, sizeof(*uart), GFP_KERNEL);
+	if (!uart)
+		return -ENOMEM;
+
 	data = devm_kzalloc(dev, sizeof(*data), GFP_KERNEL);
 	if (!data)
 		return -ENOMEM;
```

Comment:
I understand that `devm_kzalloc` will free the memory when the driver detaches, but won't this get called and allocate a copy of these for every device that is probed, which will exist in memory until the driver is unloaded (or perhaps until the device is removed).

If the probe function fails, are these allocations going to be automatically cleaned up right away? If so, then the early exist paths with errors would be ok.

In the successful case, `data` gets handed off to be used later, but `uart` is no longer needed. I think you could free `uart` at the end of the probe function since it is no longer needed.